app/src/usb_hid.c

```diff
@@ -61,4 +136,16 @@ static int zmk_usb_hid_init(const struct device *_arg) {
     return 0;
 }
 
+#if IS_ENABLED(CONFIG_ZMK_USB_BOOT)
+static int usb_hid_listener(const zmk_event_t *eh) {
+    if (as_zmk_usb_conn_state_changed(eh) && zmk_usb_get_status() == USB_DC_RESET) {
```

Comment:
I think there's a race condition here: by the time this event is delivered to the listener, the status is no longer guaranteed to be `USB_DC_RESET`, even if that was the status change that triggered the event to be sent. In practice, it doesn't work for me at all, while the previous version that set the protocol directly in `usb_status_cb` did.

Reply:
That seems like a reasonable theory. `usb_status_cb()` sets `usb_status` on some arbitrary thread before scheduling this event on the main thread. This seems bad, as `usb_status` may be accessed by different threads with no synchronization.

A couple ideas for improving this:

1. Do not directly set `usb_status` inside `usb_status_cb()`. Instead pass the status to the main thread with a [FIFO or message queue](https://docs.zephyrproject.org/3.0.0/reference/kernel/index.html#data-passing), then read it on the main thread and set `usb_status` there immediately before sending the `zmk_usb_conn_state_changed` event.
2. Add a `usb_dc_status_code` field to `struct zmk_usb_conn_state_changed` so you can query it directly from the event instead of calling `zmk_usb_get_status()` (though doing 1 would make this unnecessary to prevent a race condition).